base/reshapedarray.jl

```diff
@@ -0,0 +1,159 @@
+end
+
+function reshape(p::(AbstractArray,LinearFast), dims::Dims)
+    parent = p[1]
+    prod(dims) == length(parent) || throw(DimensionMismatch("Must have the same number of elements"))
+    ReshapedArray(parent, (IndexMD((FastDivInteger1{Int}(),), dims),), dims)
+end
+
+reshape(parent::ReshapedArray, dims::Dims) = reshape(parent.parent, dims)
+
+size(A::ReshapedArray) = A.dims
+similar{T}(A::ReshapedArray, ::Type{T}, dims::Dims) = Array(T, dims)
+linearindexing(A::ReshapedArray) = linearindexing(A.parent)
+
+size(index::IndexMD) = index.dims_view
+
+@inline function getindex(indx::IndexMD{1}, indexes::Int...)
+    sub2ind(indx.dims_view, indexes...)
+end
+@inline function getindex(indx::IndexMD, indexes::Int...)
+    ind2sub(indx.invstride_parent, sub2ind(indx.dims_view, indexes...))
+end
+
+consumes{In,Out,MI}(::Type{IndexMD{In,Out,MI}}) = Out
+consumes(::Type{Colon}) = 1
+produces{In,Out,MI}(::Type{IndexMD{In,Out,MI}}) = In
+produces(::Type{Colon}) = 1
+
+getindex(A::ReshapedArray) = A.parent[1]
+getindex(A::ReshapedArray, indx::Real) = A.parent[indx]
+
+stagedfunction getindex{T,N,P,I}(A::ReshapedArray{T,N,P,I}, indexes::Real...)
+    length(indexes) == N || throw(DimensionMismatch("Must index with all $N indexes"))
+    c = map(consumes, I)
+    breaks = [0;cumsum([c...])]
+    argbreaks = Any[]
+    for i = 1:length(c)
+        ab = Expr[]
+        for j = breaks[i]+1:breaks[i+1]
+            push!(ab, :(indexes[$j]))
+        end
+        push!(argbreaks, ab)
+    end
+    argsin = Expr[:(getindex(A.indexes[$i], $(argbreaks[i]...))) for i = 1:length(c)]
+    np = map(produces, I)
+    npc = [0;cumsum([np...])]
+    argsout = Expr[]
+    if length(argsin) > 1
+        for i = 1:npc[end]
+            j = findlast(npc .< i)
+            di = i - npc[j]
+            push!(argsout, :(tindex[$j][$di]))
+        end
+    else
+        for i = 1:npc[end]
+            push!(argsout, :(tindex[$i]))
+        end
+    end
+    meta = Expr(:meta, :inline)
+    ex = length(argsin) == 1 ?
+        quote
+            $meta
+            tindex = $(argsin...)
+            getindex(A.parent, $(argsout...))
+        end :
+        quote
+            $meta
+            tindex = tuple($(argsin...))
+            getindex(A.parent, $(argsout...))
+        end
+    ex
+end
+
+# Like strides, but operates on a Dims tuple, and returns one extra element (the total size)
+dimstrides(::()) = ()
+dimstrides(s::Dims) = dimstrides((1,), s)
+dimstrides(t::Tuple, ::()) = t
+@inline dimstrides(t::Tuple, sz::Dims) = dimstrides(tuple(t..., t[end]*sz[1]), tail(sz))
+
+ind2sub(dims::(FastDivInteger,), ind::Integer) = ind
+@inline ind2sub(dims::(FastDivInteger,FastDivInteger), ind::Integer) = begin
+    dv, rm = divrem(ind-1,dims[1])
+    rm+1, dv+1
+end
+@inline ind2sub(dims::(FastDivInteger,FastDivInteger,FastDivInteger...), ind::Integer) = begin
+    dv, rm = divrem(ind-1,dims[1])
+    tuple(rm+1, ind2sub(tail(dims),dv+1)...)
+end
+
```

Comment:
I guess that once the improved version of the generic definition of `ind2sub` is merged into base, these specialized calls should no longer be necessary, since they are essentially the same as the generic versions?